src/crimson/osd/recovery_backend.cc

```diff
@@ -161,54 +159,49 @@ seastar::future<BackfillInterval> RecoveryBackend::scan_for_backfill(
-              crimson::osd::ObjectContextRef obc;
-              if (pg.is_primary()) {
-                obc = shard_services.obc_registry.maybe_get_cached_obc(object);
-              }
-              if (obc) {
-                if (obc->obs.exists) {
-                  logger().debug("scan_for_backfill found (primary): {}  {}",
-                                 object, obc->obs.oi.version);
-                  version_map[object] = obc->obs.oi.version;
-                } else {
-                  // if the object does not exist here, it must have been removed
-                  // between the collection_list_partial and here.  This can happen
-                  // for the first item in the range, which is usually last_backfill.
-                }
-                return seastar::now();
-              } else {
-                return backend->load_metadata(object).safe_then(
-                  [&version_map, object] (auto md) {
-                    if (md->os.exists) {
-                      logger().debug("scan_for_backfill found: {}  {}",
-                                     object, md->os.oi.version);
-                      version_map[object] = md->os.oi.version;
-                    }
-                    return seastar::now();
-                  }, PGBackend::load_metadata_ertr::assert_all{});
-              }
-          }).then(
-            [&version_map, &start, next=std::move(next), this] {
-              BackfillInterval bi;
-              bi.begin = start;
-              bi.end = std::move(next);
-              bi.version = pg.get_info().last_update;
-              bi.objects = std::move(version_map);
-              logger().debug("{} BackfillInterval filled, leaving",
-                             "scan_for_backfill");
-              return seastar::make_ready_future<BackfillInterval>(std::move(bi));
-            });
-        });
+  auto version_map = seastar::make_lw_shared<std::map<hobject_t, eversion_t>>();
+  return backend->list_objects(start, max).then(
+    [this, start, version_map] (auto&& ret) {
+    auto&& [objects, next] = std::move(ret);
+    return seastar::parallel_for_each(std::move(objects),
+      [this, version_map] (const hobject_t& object) {
+      crimson::osd::ObjectContextRef obc;
+      if (pg.is_primary()) {
+        obc = shard_services.obc_registry.maybe_get_cached_obc(object);
+      }
+      if (obc) {
+        if (obc->obs.exists) {
+          logger().debug("scan_for_backfill found (primary): {}  {}",
+                         object, obc->obs.oi.version);
+          version_map->emplace(object, obc->obs.oi.version);
+        } else {
+          // if the object does not exist here, it must have been removed
+          // between the collection_list_partial and here.  This can happen
+          // for the first item in the range, which is usually last_backfill.
+        }
+        return seastar::now();
+      } else {
+        return backend->load_metadata(object).safe_then(
```

Comment:
@rzarzynski shall we acquire an exlusive lock while loading the metadata for `object`? BTW, i think we should s/object/oid/ here.

Reply:
I don't think so regarding the classical does – it just queries `ObjectStore`:

```cpp
void PrimaryLogPG::scan_range(
  int min, int max, BackfillInterval *bi,
  ThreadPool::TPHandle &handle)
{
  // ...
    } else {
      bufferlist bl;
      int r = pgbackend->objects_get_attr(*p, OI_ATTR, &bl);
      /* If the object does not exist here, it must have been removed
       * between the collection_list_partial and here.  This can happen
       * for the first item in the range, which is usually last_backfill.
       */
      if (r == -ENOENT)
        continue;

      ceph_assert(r >= 0);
      object_info_t oi(bl);
      bi->objects[*p] = oi.version;
      dout(20) << "  " << *p << " " << oi.version << dendl;
    }

```

```cpp
int PGBackend::objects_get_attr(
  const hobject_t &hoid,
  const string &attr,
  bufferlist *out)
{
  bufferptr bp;
  int r = store->getattr(
    ch,
    ghobject_t(hoid, ghobject_t::NO_GEN, get_parent()->whoami_shard().shard),
    attr.c_str(),
    bp);
  if (r >= 0 && out) {
    out->clear();
    out->push_back(std::move(bp));
  }
  return r;
}
```

>  i think we should s/object/oid/ here.

:+1:. Will make a commit just after the PTO.